### ISSUE TITLE
Encoder1 Intermediate Layer Issue (DropLastElement)

### ISSUE THREAD

**Issue by tonyma163** [narger-ef/FHE-BERT-Tiny#1]
Hi there,

I cloned this repo and started testing, but I encountered an error in the encoder1 intermediate layer.

I have checked, and the issue seems to occur when executing the ```eval_gelu_function```.

```
terminate called after throwing an instance of 'lbcrypto::OpenFHEException'
  what():  /home/tony/Github/openfhe-development/src/core/include/lattice/hal/default/dcrtpoly-impl.h:l.695:DropLastElement(): DropLastElement: Removing last element of DCRTPoly renders it invalid.
Aborted (core dumped)
```

![image](https://github.com/user-attachments/assets/ca67704b-14c5-48a4-a82a-f3f7a75f999a)

It might be due to a version difference in OpenFHE (I am using the latest one - v1.2.1), so could you please let me know which version is being used in this repo?

Thank you for your help!

**Comment by lorenzorovida** [comment]
Dear Tony, 
I've been using OpenFHE v1.0.4.

With newer versions of OpenFHE, that error may pop up. This happens because of an additional multiplication performed by the bootstrapping. I suggest you two solutions:

- Relax the security parameters to (NotSet) and increase the circuit depth by 1. This will not give you a 128-bits secure circuit anymore, something less.
- Increase the ring dimension to $2^17$ and increase the circuit depth by 1. This will still give you $>128$-bits of security, but time and memory will be doubled.

**Comment by tonyma163** [comment]
I installed the OpenFHE v1.0.4 and it works, thank you!

**Comment by tonyma163** [comment]
@narger-ef Hello sir, may I ask how did you calculate the degrees?

**Comment by lorenzorovida** [comment]
> @narger-ef Hello sir, may I ask how did you calculate the degrees?

https://github.com/openfheorg/openfhe-development/blob/main/src/pke/examples/FUNCTION_EVALUATION.md

**Comment by tonyma163** [comment]
Maybe I asked the question unclearly.

How did you determine the suitable polynomial degree for different intervals and functions? What level of approximation error did you consider acceptable?

**Comment by lorenzorovida** [comment]
> Maybe I asked the question unclearly.
> 
> How did you determine the suitable polynomial degree for different intervals and functions? What level of approximation error did you consider acceptable?

No prob!

Usually, higher degrees mean better precision (not always true though). Nevertheless, a balance between circuit depth and precision must be found... Tbh, the degree was found experimentally, but can surely be topic of further analysis.

When first designing the circuit, we used toy parameters and very precise approximations, then we relaxed the degrees when increasing the security level.

If you have further questions, feel free to ask (maybe in another issue)! :-)
